File: Data Structures & Algorithms/maximum-product-subarray/submission-7.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        
        maxProd, minProd = nums[0], nums[0]
        res = maxProd
        for i in range(1, len(nums)):
            tmp = maxProd * nums[i]
            maxProd = max(maxProd * nums[i], nums[i], minProd * nums[i])
            minProd = min(tmp, nums[i], minProd * nums[i])
            res = max(res, maxProd)

        return res

File: Data Structures & Algorithms/maximum-sum-circular-subarray/submission-7.cpp
class Solution {
public:
    int maxSubarraySumCircular(vector<int>& nums) {
        int maxSum = nums[0], currMaxSum = 0;
        int minSum = nums[0], currMinSum = 0;
        int sum = 0;
        for (int num: nums) {

            sum += num;
            
            currMaxSum += num;
            currMinSum += num;
            
            maxSum = max(maxSum, currMaxSum);
            minSum = min(minSum, currMinSum);

            if (currMaxSum < 0) currMaxSum = 0;
            if (currMinSum > 0) currMinSum = 0;
        }

        return maxSum > 0 ? max(maxSum, sum - minSum) : maxSum;
    }
};